jupyterhub/crypto.py

```diff
@@ -0,0 +1,161 @@
+
+import base64
+from binascii import a2b_hex
+from concurrent.futures import ThreadPoolExecutor
+import json
+import os
+
+from traitlets.config import SingletonConfigurable, Config
+from traitlets import (
+    Any, Dict, Integer, List,
+    default, observe, validate,
+)
+
+try:
+    import cryptography
+    from cryptography.fernet import Fernet, MultiFernet, InvalidToken
+except ImportError:
+    cryptography = None
+    class InvalidToken(Exception):
+        pass
+
+
+KEY_ENV = 'JUPYTERHUB_CRYPT_KEY'
+
+class EncryptionUnavailable(Exception):
+    pass
+
+class CryptographyUnavailable(EncryptionUnavailable):
+    def __str__(self):
+        return "cryptography library is required for encryption"
+
+class NoEncryptionKeys(EncryptionUnavailable):
+    def __str__(self):
+        return "Encryption keys must be specified in %s env" % KEY_ENV
+
+
+def _validate_key(key):
+    """Validate and return a 32B key
+
+    Args:
+    key (bytes): The key to be validated.
+        Can be:
+        - base64-encoded (44 bytes)
+        - hex-encoded (64 bytes)
+        - raw 32 byte key
+
+    Returns:
+    key (bytes): raw 32B key
+    """
+    if isinstance(key, str):
+        key = key.encode('ascii')
+
+    if len(key) == 44:
+        try:
+            key = base64.urlsafe_b64decode(key)
+        except ValueError:
+            pass
+
+    elif len(key) == 64:
+        try:
+            # 64B could be 32B, hex-encoded
+            return a2b_hex(key)
+        except ValueError:
+            # not 32B hex
+            pass
+
+    if len(key) != 32:
+        raise ValueError("Encryption keys must be 32 bytes, hex or base64-encoded.")
+
+    return key
+
+class CryptKeeper(SingletonConfigurable):
+    """Encapsulate encryption configuration
+
+    Use via the encryption_config singleton below.
+    """
+
+    n_threads = Integer(max(os.cpu_count(), 1), config=True,
```

Comment:
I've been thinking we should base almost all our threadpools on concurrent_spawn_limit - so we don't end up blocking waiting on any single thread. Then we can set the concurrent_spawn_limit to a reasonable default, and have it propogate forward. It lets admins also tweak one variable that is easy to understand, rather than many related ones. For example, if n_threaads is 1 we're effectively limiting the number of concurrent user logins to 1, regardless of any other setting - I don't think you'd ever want to limit login to 1 but spawn to 50, for example.